subprojects/provider/src/main/kotlin/org/gradle/kotlin/dsl/execution/Interpreter.kt

```diff
@@ -0,0 +1,559 @@
+                        } ?: targetScopeClassPath
+
+                    scriptHost.scriptSource.withLocationAwareExceptionHandling {
+                        ResidualProgramCompiler(
+                            outputDir,
+                            compilationClassPath,
+                            sourceHash,
+                            programKind,
+                            programTarget,
+                            host.implicitImports
+                        ).emitStage2ProgramFor(
+                            File(scriptPath),
+                            originalScriptPath
+                        )
+                    }
+                }
+
+            logClassLoadingOf(scriptTemplateId, scriptHost.scriptSource)
+
+            return loadClassInChildScopeOf(
+                targetScope,
+                originalScriptPath,
+                cacheDir,
+                scriptTemplateId,
+                accessorsClassPath)
+        }
+
+        open fun eval(specializedProgram: Class<*>, scriptHost: KotlinScriptHost<*>) {
+            withContextClassLoader(specializedProgram.classLoader) {
+                instantiate(specializedProgram)
+                    .execute(this, scriptHost)
+            }
+        }
+
+        private
+        fun instantiate(specializedProgram: Class<*>) =
+            specializedProgram.getDeclaredConstructor().newInstance() as ExecutableProgram
+    }
+}
+
+
+internal
+fun templateIdFor(programTarget: ProgramTarget, programKind: ProgramKind, stage: String): String =
+    programTarget.name + "/" + programKind.name + "/" + stage
+
+
+internal
+fun classLoaderScopeIdFor(scriptPath: String, stage: String) =
+    "kotlin-dsl:$scriptPath:$stage"
+
+
+internal
+fun locationAwareExceptionHandlingFor(e: Throwable, scriptClass: Class<*>, scriptSource: ScriptSource): Nothing {
+    val targetException = maybeUnwrapInvocationTargetException(e)
+    val locationAware = locationAwareExceptionFor(targetException, scriptClass, scriptSource)
+    throw locationAware ?: targetException
+}
+
+
+private
+fun locationAwareExceptionFor(
+    original: Throwable,
+    scriptClass: Class<*>,
+    scriptSource: ScriptSource
+): LocationAwareException? {
+
+    val scriptClassName = scriptClass.name
+    val scriptClassNameInnerPrefix = "$scriptClassName$"
+
+    fun scriptStackTraceElement(element: StackTraceElement) =
+        element.className?.run {
+            equals(scriptClassName) || startsWith(scriptClassNameInnerPrefix)
+        } == true
+
+    tailrec fun inferLocationFrom(exception: Throwable): LocationAwareException? {
+
+        if (exception is LocationAwareException) {
+            return exception
+        }
+
+        exception.stackTrace.find(::scriptStackTraceElement)?.run {
+            return LocationAwareException(original, scriptSource, lineNumber.takeIf { it >= 0 })
+        }
+
+        val cause = exception.cause ?: return null
+        return inferLocationFrom(cause)
+    }
```

Comment:
So I understand _what_ this is doing, but I don't understand _why_ you are doing it. Can you add a bit of a comment explaining?